File: Estruturais/Decorator/Decorator.js
 // Componente notificador
class ComponenteNotificador{
    enviar(msg){
        console.log(`Enviando Msg: ${msg}`)
    }
}

// componente Decorador Base:

class DecoradorNotificador{
    constructor(componente){
        this.componente = componente
    }

    enviar(msg){
        this.componente.enviar(msg)
    }
}

// Compoentes decorator Concreto 1 - SMS
class NotificadorSMS extends DecoradorNotificador{
    enviar(msg){
        super.enviar(msg);
        console.log(`Enviando Msg por SMS: ${msg}`)
    }
}

// Componentes Decoraor Concreto 2 - Whatsapp

class NotificadorWhatsapp extends DecoradorNotificador{
    enviar(msg){
        super.enviar(msg)
        console.log(`Enviando Msg por Whatsapp: ${msg}`)
    }
}

class NotificadorEmail extends DecoradorNotificador{
    enviar(msg){
        super.enviar(msg)
        console.log(`Enviando Msg por Email: ${msg}`)
    }
}

// USO PELO CLIENTE:
const notificador = new ComponenteNotificador()
const notificadorSMS = new NotificadorSMS(notificador)
const notificadorWhatsapp = new NotificadorWhatsapp(notificadorSMS)
notificadorWhatsapp.enviar("Pedido do Habbib's saindo...")

File: Estruturais/Decorator/Decorator.py
class ComponenteNotificador:
    def enviar(self, msg):
        print(f'Enviando msg: {msg}')

class DecoradorNotificador:
    def __init__(self, componente: ComponenteNotificador):
        self.componente = componente

    def enviar(self, msg):
        self.componente.enviar(msg)

class NotificadorSMS(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando Msg por SMS: {msg}')

class NotificadorWhatsapp(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando Msg por whatsapp: {msg}')

class NotificadorEmail(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando Msg por Email: {msg}')

notificador = ComponenteNotificador()
notificadorSMS = NotificadorSMS(notificador)
notificadorWhatsapp = NotificadorWhatsapp(notificadorSMS)
notificadorWhatsapp.enviar("Pedido do Habbib's saindo...")